File: python/atelier_bench/exporters/xml_exporter.py
"""xml exporter."""

from __future__ import annotations

from collections.abc import Sequence

from ..contracts.report_contract import ReportRow
from ..support.abstract_exporter import AbstractExporter


class XmlExporter(AbstractExporter):
    def __init__(self) -> None:
        super().__init__("xml", "application/xml")

    def export(self, rows: Sequence[ReportRow]) -> str:
        body = "".join(f'<row label="{row.label}" cents="{row.cents}"/>' for row in rows)
        return f"<rows>{body}</rows>"

File: go/atelier/exporters/xml_exporter.go
// Package exporters carries the 8 exporters. Each embeds support.BaseExporter
// (Extension, FormatCell), declares only Export, and registers itself under
// its extension in an init function.
package exporters

import (
	"fmt"
	"strings"

	"atelier.example/lane/atelier/contracts"
	"atelier.example/lane/atelier/support"
)

// Compile-time proof of satisfaction.
var _ contracts.ExporterContract = (*XmlExporter)(nil)

// XmlExporter renders report rows as xml.
type XmlExporter struct {
	support.BaseExporter
	MimeType string
}

// NewXmlExporter builds the exporter.
func NewXmlExporter() *XmlExporter {
	return &XmlExporter{BaseExporter: support.NewBaseExporter("xml"), MimeType: "application/xml"}
}

// Export is the only contract method this type declares.
func (e *XmlExporter) Export(rows []contracts.ReportRow) string {
	var body strings.Builder
	for _, row := range rows {
		body.WriteString(fmt.Sprintf("<row label=%q cents=%q/>", row.Label, fmt.Sprint(row.Cents)))
	}
	return fmt.Sprintf("<rows>%s</rows>", body.String())
}

// init registers this exporter under its extension.
func init() {
	Register(NewXmlExporter())
}
